src/classes/RD2_ScheduleService.cls

```diff
@@ -185,6 +193,152 @@ public without sharing class RD2_ScheduleService {
+            currentDate = visualizationStartDate;
+            Integer dateCount = 0;
+            List <RecurringDonationSchedule__c> filteredSchedules;
+
+            if (schedulesByRDIds.get(rd.Id) == null) {
+                continue;
+            }
+
+            while (dateCount < maxDates) {
+
+                // We are changing currentDate with each loop, so the active schedules potentially change with each loop
+                filteredSchedules = new List<RecurringDonationSchedule__c>();
+                for (RecurringDonationSchedule__c schedule : schedulesByRDIds.get(rd.Id)) {
+                    if (schedule.EndDate__c == null || schedule.EndDate__c >= currentDate) {
+                        filteredSchedules.add(schedule);
+                    }
+                }
+
+                ins = getNextDonationSchedule(filteredSchedules);
+                if (ins == null) {
+                    currentDate = date.today();
+                    break;
+                }
+
+                vizSchedules.add(ins);
+                dateCount++;
+                currentDate = ins.nextDonationDate.addDays(1);
+            }
+
+            installmentsByRDIds.get(rd.Id).addAll(vizSchedules);
+
+        }
+
+        currentDate = date.today();
+
+        return installmentsByRDIds;
+    }
+
+    /***
+    * @description Wrapper class for Recurring Donation schedule data
+    */
+    public class Installment {
+
+        public Id       campaignId {get; set;}
+        public String   campaignName {get; set;}
+        public Date     endDate {get; set;}
+        public String   externalId {get; set;}
+        public Decimal  installmentAmount {get; set;}
+        public Date     nextDonationDate {get; set;}
+        public String   paymentMethod {get; set;}
+        public Id       recurringDonationId {get; set;}
+        public String   recurringDonationName {get; set;}
+        public Id       scheduleId {get; set;}
+        public Date     startDate {get; set;}
+
+        Installment(RecurringDonationSchedule__c rdSchedule) {
+            campaignId =            rdSchedule.Campaign__c;
+            campaignName =          rdSchedule.Campaign__r.Name;
+            endDate =               rdSchedule.EndDate__c;
+            externalId =            rdSchedule.ExternalId__c;
+            installmentAmount =     rdSchedule.InstallmentAmount__c;
+            paymentMethod =         rdSchedule.PaymentMethod__c;
+            recurringDonationId =   rdSchedule.RecurringDonation__c;
+            recurringDonationName = rdSchedule.RecurringDonation__r.Name;
+            scheduleId =            rdSchedule.Id;
+            startDate =             rdSchedule.StartDate__c;
+        }
+    }
+
+    /***
+    * @description Calculates the date of the next donation
+    * @param rdSchedules Recurring Donation Schedule records
+    * @return Date
+    */
+    public Installment getNextDonationSchedule(List<RecurringDonationSchedule__c> rdSchedules) {
```

Comment:
Rather than define `currentDate` in the calling method, I'm suggesting to pass that value as a parameter to this method and have this method "override" `currentDate`. This way the logic related to that is encapsulated in the method it's used from. Also, does this need to be a public method?

```suggestion
    public Installment getNextInstallment(List<RecurringDonationSchedule__c> rdSchedules, Date startingDate) {
         currentDate = startingDate;
```

Reply:
I'll use startDate instead of startingDate to be consistent with what we have done elsewhere.

Reply:
Actually, having already worked through this once and watched how it borked everything, I'm going to push back a bit on this one.  While I agree that encapsulating changes is generally a good practice, it doesn't really buy us anything in this case.  Now that we have bulkified the caller, we have to reset to startDate with every iteration of the schedule analysis for each RD passed in.  We also need to reset to current date at the end of the entire process to ensure that any caller holding a reference to RD2_ScheduleService will end up with a status that is what they rightfully expect.  To me, that means that encapsulating startDate as you suggest is actually a negative, not a positive.  It doesn't get us anything of value and makes it more cumbersome to do what we require.

Reply:
Alternatively....  is it worth considering a more radical set of changes to accept date inputs pretty much everywhere?  The fact that we are relying on a class variable to handle currentDate is useful for testing, but is resulting in some awkward patterns like what we are encountering here.  Worth some refactoring perhaps?